odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2405 @@
+        """Implement ``self[indices] = values``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be written to.
+        values : scalar, array-like or `NumpyTensor`
+            The value(s) that are to be assigned.
+
+            If ``index`` is an integer, ``value`` must be a scalar.
+
+            If ``index`` is a slice or a sequence of slices, ``value``
+            must be broadcastable to the shape of the slice.
+
+        Examples
+        --------
+        For 1d spaces, entries can be set with scalars or sequences of
+        correct shape:
+
+        >>> space = odl.rn(3)
+        >>> x = space.element([1, 2, 3])
+        >>> x[0] = -1
+        >>> x[1:] = (0, 1)
+        >>> x
+        rn(3).element([-1.,  0.,  1.])
+
+        It is also possible to use tensors of other spaces for
+        casting and assignment:
+
+        >>> space = odl.rn((2, 3))
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> x[0, 1] = -1
+        >>> x
+        rn((2, 3)).element(
+            [[ 1., -1.,  3.],
+             [ 4.,  5.,  6.]]
+        )
+        >>> short_space = odl.tensor_space((2, 2), dtype='short')
+        >>> y = short_space.element([[-1, 2],
+        ...                          [0, 0]])
+        >>> x[:, :2] = y
+        >>> x
+        rn((2, 3)).element(
+            [[-1.,  2.,  3.],
+             [ 0.,  0.,  6.]]
+        )
+
+        The Numpy assignment and broadcasting rules apply:
+
+        >>> x[:] = np.array([[0, 0, 0],
+        ...                  [1, 1, 1]])
+        >>> x
+        rn((2, 3)).element(
+            [[ 0.,  0.,  0.],
+             [ 1.,  1.,  1.]]
+        )
+        >>> x[:, 1:] = [7, 8]
+        >>> x
+        rn((2, 3)).element(
+            [[ 0.,  7.,  8.],
+             [ 1.,  7.,  8.]]
+        )
+        >>> x[:, ::2] = -2.
+        >>> x
+        rn((2, 3)).element(
+            [[-2.,  7., -2.],
+             [-2.,  7., -2.]]
+        )
+        """
+        if isinstance(values, NumpyTensor):
+            self.data[indices] = values.data
+        else:
+            self.data[indices] = values
+
+    @property
+    def real(self):
+        """Real part of ``self``.
+
+        Returns
+        -------
+        real : `NumpyTensor`
+            Real part of this element as a member of a
+            `NumpyTensorSpace` with corresponding real data type.
+
+        Examples
+        --------
+        >>> space = odl.cn(3)
+        >>> x = space.element([1 + 1j, 2, 3 - 3j])
+        >>> x.real
+        rn(3).element([ 1.,  2.,  3.])
+        """
+        if self.space.is_real_space:
+            return self
+        elif self.space.is_complex_space:
+            # Definitely non-contiguous
+            real_space = self.space.astype(self.space.real_dtype, order='A')
```

Comment:
I now remember our discussion on this, but add a comment to clarify

Reply:
We need to set `order='A'` because the array will be discontiguous.
Alternatively we could always use `order='A'` when using `.real` on a complex space. But I would find that counter-intuitive, I'd expect the order to be preserved in that operation **on spaces**.